gpii/node_modules/preferencesServer/test/prefsSafesTests.js

```diff
@@ -0,0 +1,715 @@
+    var expectedStatusCode = fixture.expectedStatusCode || 200;
+    return {
+        name: fixture.name,
+        expect: 2,
+        components: {
+            getRequest: {
+                type: "kettle.test.request.http",
+                options: {
+                    path: "/prefssafe/" + fixture.url,
+                    method: "GET",
+                    port: 8081,
+                    termMap: {
+                        prefsSafeId: fixture.prefsSafeId
+                    }
+                }
+            }
+        },
+        sequence: [{
+            func: "{getRequest}.send"
+        }, {
+            event: "{getRequest}.events.onComplete",
+            listener: "gpii.tests.preferencesServer.testResponse",
+            args: ["{arguments}.0", fixture.expected, "{getRequest}.nativeResponse.statusCode", expectedStatusCode]
+        }]
+    };
+};
+
+gpii.tests.preferencesServer.prefsSafes.get.successFixtures = [
+    {
+        name: "GET: Basic retrieval of a prefs safe",
+        prefsSafeId: "prefsSafe-7",
+        url: "%prefsSafeId",
+        expected: {
+            "id": "prefsSafe-7",
+            "type": "prefsSafe",
+            "schemaVersion": "0.1",
+            "prefsSafeType": "user",
+            "name": null,
+            "email": null,
+            "preferences": {
+                "flat": {
+                    "name": "bit of stuff",
+                    "contexts": {
+                        "gpii-default": {
+                            "name": "Default preferences",
+                            "preferences": {
+                                "http://registry.gpii.net/common/onScreenKeyboard/enabled": true
+                            },
+                            "metadata": []
+                        }
+                    },
+                    "metadata": []
+                }
+            },
+            "timestampCreated": "2017-12-14T19:55:11.641Z",
+            "timestampUpdated": null
+        }
+    }
+];
+
+gpii.tests.preferencesServer.prefsSafes.get.failFixtures = [
+    {
+        name: "GET error: Fetching a safe that does not exist returns an error payload.",
+        prefsSafeId: "prefsSafe-2999",
+        url: "%prefsSafeId",
+        expectedStatusCode: 404,
+        expected: {
+            "isError": true,
+            "message": "Cannot lookup prefsSafe"
+        }
+    }
+];
+
+/////////////////////// TESTING prefsSafes with keys GET ////////////////////////////
+fluid.registerNamespace("gpii.tests.preferencesServer.prefsSafes.getSafeWithKeys");
+
+gpii.tests.preferencesServer.prefsSafes.getSafeWithKeys.buildTestDef = function (fixture) {
+    var expectedStatusCode = fixture.expectedStatusCode || 200;
+    return {
+        name: fixture.name,
+        expect: 2,
+        components: {
+            getRequest: {
+                type: "kettle.test.request.http",
+                options: {
+                    path: "/prefssafe-with-keys/" + fixture.url,
```

Comment:
Can you write a test for this endpoint that a preferences safe only associates with one gpii key rather than more than one keys. I'm getting this error in this case:
```
FATAL ERROR: Uncaught exception: Cannot read property 'rows' of undefined
```

Reply:
Actually, even if a preferences safe associates with more than one "gpiiKey" documents (without having any "gpiiCloudSafeCredential" documents), the same error still occurs.

Reply:
Ok, the `FATAL ERROR` is definately something to be fixed, and you noted it below as well. However, I do believe that with the internal API it should be possible to associate more than one gpiiKey with a preferences set, since we have the ability to revoke and create new ones, and theoretically support having more than one key to a prefssafe.

Reply:
Sorry that my original point was not clear enough.

I totally agree that one prefs safe can associate with multiple GPII keys. What I meant was, when I tested this endpoint, when a prefs safe associated with one or more GPII keys (in the database) without associating with any login credentials, the error mentioned above was hit.